docs/specs/tsd-008-probe-on-startup.md

```diff
@@ -0,0 +1,88 @@
+
+More background discussion on this option, including other possible avenues, can 
+be viewed [here](https://github.com/influxdata/telegraf/issues/16028).
+
+## Probing
+
+First, we must define what it means for a plugin to `Probe`. Probing is an 
+action whereby the plugin will attempt to communicate with its external service, 
+device, entity, or executable as if it were gathering real metrics. The payload 
+of this communication attempt is not used to record any metrics. 
+
+Plugins that support probing must implement the `ProbePlugin` interface. Such 
+plugins should behave in the following manner:
+
+1. Return an error if the external dependencies (hardware, services, 
+executables, etc.) of the plugin are not available.
+2. Return an error if information cannot be gathered (in the case of inputs) or 
+sent (in the case of outputs) due to unrecoverable issues. For example, invalid 
+authentication, missing permissions, or non-existent endpoints.
+3. Otherwise, return `nil`.
+
+## Configuration
+
+The Telegraf project has already introduced the `startup_error_behavior` 
+configuration which allows the user to define how Telegraf should behave if the 
+plugin fails to start. The `ignore` value allows the plugin to be ignored if the 
+`Start()`/`Connect()` method returns an error. We propose to add an additional 
+value to this parameter called `probe`, that will behave as a superset of the 
+`ignore` behavior. When `startup_error_behavior=probe`, Telegraf will perform 
+the following steps on plugin startup:
+
+1. Check if the plugin implements `ProbePlugin`. If it does not, Telegraf will 
+fatally exit with a log message indicating that the supplied configuration is 
+invalid for the plugin.
+2. Call the `Start()` or `Connect()` method.
+3. If an error is returned, cause the plugin to be ignored in the same manner as 
+if `ignore` was specified.
+4. Call `Probe()`. If `Probe()` returns an error, cause the plugin to be ignored.
+
+## Plugin Requirements
+
+As already stated, plugins participating in the `probe` scheme must implement 
+the `ProbePlugin` interface. The exact way the plugin implements the behavior 
+will depend on the plugin in question, but generally it should take the same 
+actions as it would with `Gather()` or `Write()`, such that failures during 
+`Gather()`/`Write()` would also imply a failure of `Probe()`. If `Probe()` 
+returns an error, the plugin's `Close()` method should still be safe to call.
+
+It should be noted that for output plugins, it's advisable to implement 
+`Probe()` in a way that doesn't write real metrics to the backend if possible. 
+How this might be done depends on the output in question.
+
+Plugins should return `nil` upon successful probes or a retriable error (via 
+predefined error type) to enable the defined ignoring behavior. A non-retriable 
+error (via predefined error type) or a generic error will bypass the startup 
+error behaviors and Telegraf must fail and exit in the startup phase.
```

Comment:
```suggestion
Plugins that allow probing must implement  the `ProbePlugin` interface. The
exact implementation depend on the plugin's functionality and requirements,
but generally it should take the same actions as it would during normal operation
e.g. calling `Gather()` or `Write()` and check if errors occur. If probing fails, it must
be safe to call the plugin's `Close()` method.

Input plugins must *not* produce metrics, output plugins must *not* send any
metrics to the service. Plugins must *not* influence the later data processing or
collection by modifying the internal state of the plugin or the external state of the
service or hardware. For example, file-offsets or other service states must be
reset to not loose data during the first gather or write cycle.

Plugins must return `nil` upon successful probing or an error otherwise.
```